texar/modules/connectors/connectors.py

```diff
@@ -0,0 +1,786 @@
+                 distribution_kwargs: Optional[Dict[str, Any]] = None,):
+        ConnectorBase.__init__(self, output_size, hparams)
+        if distribution_kwargs is None:
+            distribution_kwargs = {}
+        self._dstr = utils.check_or_get_instance(
+            distribution, distribution_kwargs,
+            ["torch.distributions", "torch.distributions.multivariate_normal",
+             "texar.custom"])
+
+    @staticmethod
+    def default_hparams() -> dict:
+        r"""Returns a dictionary of hyperparameters with default values.
+
+        .. code-block:: python
+
+            {
+                "activation_fn": "texar.core.layers.identity",
+                "name": "reparameterized_stochastic_connector"
+            }
+
+        Here:
+
+        "activation_fn" : str
+            The activation function applied to the outputs of the MLP
+            transformation layer. Can be a function,
+            or its name or module path.
+        "name" : str
+            Name of the connector.
+        """
+        return {
+            "activation_fn": "texar.core.layers.identity",
+            "name": "reparameterized_stochastic_connector"
+        }
+
+    def forward(self, # type: ignore
+                transform: bool = True,
+                num_samples: Optional[Union[int, torch.Tensor]] = None
+                ) -> Tuple[Any, Any]:
+        r"""Samples from a distribution and optionally performs transformation
+        with an MLP layer.
+        The distribution must be reparameterizable, i.e.,
+        :python:`distribution.reparameterization_type = FULLY_REPARAMETERIZED`.
+
+        Args:
+            transform (bool): Whether to perform MLP transformation of the
+                distribution samples. If ``False``, the structure/shape of a
+                sample must match :attr:`output_size`.
+            num_samples (optional): An ``int`` or ``int`` ``Tensor``.
+                Number of samples to generate. If not given,
+                generate a single sample. Note that if batch size has
+                already been included in :attr:`distribution`'s dimensionality,
+                :attr:`num_samples` should be left as ``None``.
+
+        Returns:
+            A ``tuple`` (output, sample), where
+            - output: A ``Tensor`` or a (nested) ``tuple`` of Tensors with
+              the same structure and size of :attr:`output_size`.
+              The batch dimension equals :attr:`num_samples` if specified,
+              or is determined by the distribution dimensionality.
+            - sample: The sample from the distribution, prior to transformation.
+        Raises:
+            ValueError: If distribution cannot be reparametrized.
+            ValueError: The output does not match :attr:`output_size`.
+        """
+
+        if num_samples:
+            sample = self._dstr.sample([num_samples]) # type: ignore
+        else:
+            sample = self._dstr.sample() # type: ignore
```

Comment:
Also, we should use `rsample` instead of `sample`.

Reply:
Thanks! Checked.